modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java

```diff
@@ -0,0 +1,388 @@
+
+                            if (crd.isLocal())
+                                initCoordinator(p, discoCache.version());
+
+                            if (!ctx.clientNode())
+                                p.resFut.listen(f -> sendSingleSingleMessage(p, crd));
+                        }
+                        else
+                            onAllServersLeft();
+                    }
+                    else if (ctx.localNodeId().equals(p.crdId)) {
+                        boolean rmvd, isEmpty;
+
+                        synchronized (mux) {
+                            rmvd = p.remaining.remove(leftNodeId);
+
+                            isEmpty = p.remaining.isEmpty();
+                        }
+
+                        if (rmvd) {
+                            assert !p.singleMsgs.containsKey(leftNodeId);
+
+                            if (isEmpty)
+                                finishProcess(p);
+                        }
+                    }
+                });
+            }
+        }, EVT_NODE_FAILED, EVT_NODE_LEFT);
+    }
+
+    /**
+     * Starts distributed process.
+     *
+     * @param req Initial request.
+     * @param id Process id.
+     */
+    public void start(UUID id, I req) {
+        try {
+            InitMessage msg = new InitMessage<>(id, type, req);
+
+            ctx.discovery().sendCustomEvent(msg);
+        }
+        catch (IgniteCheckedException e) {
+            log.warning("Unable to start process.", e);
+        }
+    }
+
+    /**
+     * Initiates process coordinator.
+     *
+     * @param p Process.
+     * @param topVer Topology version.
+     */
+    private void initCoordinator(Process p, AffinityTopologyVersion topVer) {
+        Set<UUID> aliveSrvNodesIds = new HashSet<>();
+
+        for (ClusterNode node : ctx.discovery().serverNodes(topVer)) {
+            if (ctx.discovery().alive(node))
+                aliveSrvNodesIds.add(node.id());
+        }
+
+        synchronized (mux) {
+            if (p.initCrdFut.isDone())
+                return;
+
+            assert p.remaining.isEmpty();
+
+            p.remaining.addAll(aliveSrvNodesIds);
+
+            p.initCrdFut.onDone();
+        }
+    }
+
+    /**
+     * Sends single node message to coordinator.
+     *
+     * @param p Process.
+     * @param crd Coordinator node to send message.
+     */
+    private void sendSingleSingleMessage(Process p, ClusterNode crd) {
+        assert p.resFut.isDone();
+
+        SingleNodeMessage<R> singleMsg = new SingleNodeMessage<>(p.id, type, p.resFut.result(),
+            (Exception)p.resFut.error());
+
+        if (crd.isLocal())
+            onSingleNodeMessageReceived(singleMsg, crd.id());
+        else {
+            try {
+                ctx.io().sendToGridTopic(crd, GridTopic.TOPIC_DISTRIBUTED_PROCESS, singleMsg, SYSTEM_POOL);
+            }
+            catch (IgniteCheckedException e) {
+                log.warning("Unable to send message.", e);
```

Comment:
We can't ignore this exception.
Otherwise, the process can never end.

Reply:
Done. I have added failure handler call.